NEXT-12791 - [NEW] Add responsive behavior (#196)
shopware/docs: guides/plugins/plugins/administration/adding-responsive-behavior.md

# Adding responsive behavior

> 

Reviewer: ```suggestion

## Overview
```

The Shopware 6 Administration provides two ways of adding classes to elements based on their size, the device helper and the `v-responsive` directive.
Alternatively you can use `sccs` media queries to make your plugin responsive.

> Learn how to use `sccs` here, [Add custom styles](./add-custom-styles.md).

Reviewer: `page-ref` nutzen?

```suggestion
Learn how to use `scss` here:

{% page-ref page="./add-custom-styles.md" %}
```

> ## The device helper

Reviewer: ```suggestion
## DeviceHelper
```

The DeviceHelper provides methods to get device and browser information like the current viewport size.
The helper methods can be accessed with "this.$device" in every Vue component,
since it is bound to the Vue prototype.

It provides functionality to run a function on resize to for example add classes to a element.
The example below shows you how to use the `$device.onResize` helper.

```javascript
const listener = function (ev) {
    // do something on resize with the event, like passing  
};
const scope = this;
const component = 'sw-basic-example';

this.$device.onResize({ listener, scope, component });
this.$device.removeResizeListener(component);
```

> It also provides many helper functions to get for example the screen dimensions.

Reviewer: ```suggestion
It also provides many helper functions e.g. to get the screen dimensions.
```

> Although there are many more as seen below:

Reviewer: Can or should we have a reference article about that? //CC: @PaddyS

Author: I really don't know what this refernce would be about

Reviewer: I would explain the directives you list right after this line: 

```javascript
this.$device.getViewportWidth();
this.$device.getViewportHeight()
this.$device.getDevicePixelRatio()
this.$device.getScreenWidth()
this.$device.getScreenHeight()
this.$device.getScreenOrientation()
```

Reviewer 2: I like.

```javascript
this.$device.getViewportWidth();
this.$device.getViewportHeight()

this.$device.getDevicePixelRatio()

this.$device.getScreenWidth()
this.$device.getScreenHeight()
this.$device.getScreenOrientation()
```

## v-responsive directive

The `v-responsive` directive can be used to dynamically apply classes based on an elements dimensions.

```html
<input v-responsive="{ 'is--compact': el => el.width <= 1620, timeout: 200 }">
```

> Explanation:

Reviewer: ```suggestion
Let's do a small explanation of this directive:
```

> - Apply class (in this case: 'is--compact') when the width of the element is smaller than 1620px.

Reviewer: ```suggestion
- Apply class (in this case: `is--compact`) when the width of the element is smaller than 1620px.
```

> - timeout: Sets the duration on how much the throttle should wait.

Reviewer: ```suggestion
- `timeout`: Sets the duration on how much the throttle should wait.
```

# Adding responsive behavior

## Overview
The Shopware 6 Administration provides two ways of adding classes to elements based on their size, the device helper and the `v-responsive` directive.

> Alternatively you can use `sccs` media queries to make your plugin responsive.

Reviewer 3: `sccs` -> `scss`

How are scss media queries different css media queries?

Reviewer 4: There is no difference at all. I'd suggest you refer to `CSS` queries here

Learn how to use `scss` here:

{% page-ref page="./add-custom-styles.md" %}

## DeviceHelper

The DeviceHelper provides methods to get device and browser information like the current viewport size.

> The helper methods can be accessed with "this.$device" in every Vue component,

Reviewer: ```suggestion
The helper methods can be accessed with `this.$device` in every Vue component,
```

since it is bound to the Vue prototype.

> It provides functionality to run a function on resize to for example add classes to a element.

Reviewer 4: Please revisit this sentence. That's not proper English and I find it hard to suggest anything here, as I do not understand, what you want to say here

The example below shows you how to use the `$device.onResize` helper.

```javascript
const listener = function (ev) {

>     // do something on resize with the event, like passing  

Reviewer 4: Passing ... ?

};
const scope = this;
const component = 'sw-basic-example';

this.$device.onResize({ listener, scope, component });

> this.$device.removeResizeListener(component);

Reviewer 4: Why do you remove the listener afterwards ? Please document this

```
It also provides many helper functions e.g. to get the screen dimensions.
Although there are many more as seen below:

| Function                               | Description                 |
|----------------------------------------|-----------------------------|

> | `this.$device.getViewportWidth();`     | gets the viewport width     |

Reviewer: Please use capital letters at the beginning of the sentences in the second column

| `this.$device.getViewportHeight();`    | gets the viewport height    |
| `this.$device.getDevicePixelRatio();`  | gets the device pixel ratio |
| `this.$device.getScreenWidth();`       | gets the screen width       |
| `this.$device.getScreenHeight();`      | gets screen height          |
| `this.$device.getScreenOrientation();` | gets the screen orientation |

## v-responsive directive

> The `v-responsive` directive can be used to dynamically apply classes based on an elements dimensions.

Reviewer: ```suggestion
The `v-responsive` directive can be used to dynamically apply classes based on an element's dimensions.
```
